ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+  # Both signal and query handlers can be defined as Symbols (when the name matches the
+  # method) or overriden when the name is different (or cannot be used as a method name)
+  signal :add
+  signal :subtract, name: 'my.signal.subtract'
+  
+  query :value # can use a getter here
+  query :get_previous_value, name: 'my.query.previous_value'
+
+  # It is preferable in this case to initialize the values before the execution to
+  # support starting a workflow with a signal
+  def initialize
+    @value = 0
+    @previous_value = 0
+  end
+
+  def execute
+    async.await_for { @value >= 42 }
+  end
+
+  private
+  
+  attr_reader :value
+
+  def add(val)
+    @previous_value = @value
+    @value += val
+  end
+
+  def subtract(val)
+    @previous_value = @value
+    @value -= val
+  end
+
+  def get_previous_value
+    @previous_value
+  end
+end
+```
+
+#### Alternative approach
+
+To provide extra flexibility in handling signals and queries (especially from external classes and
+methods) we will allow to define signals and queries from the workflow context:
+
+```ruby
+class CalculatorWorkflow < Temporal::Workflow
+  def execute
+    value = 0
+
+    workflow.on_signal(:add) do |val|
+      value += val
+    end
+
+    workflow.on_signal('my.signal.subtract') do |val|
+      value -= val
+    end
+
+    workflow.on_query(:value) do
+      value
+    end
+
+    # A condition block will pause until the enclosed statement returns true
+    async.await_for { value >= 42 }
+  end
+end
+```
+
+
+## Signal and query a workflow
+
+```ruby
+require 'calculator_workflow'
+
+connection = Temporal::Connection.connect('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+handle.signal(:add, 34)
+handle.query(:get_value) # => 34
+handle.signal(:subtract, 4) # This will send a 'my.signal.subtract' signal as defined
+handle.query(:get_value) # => 30
+handle.signal(:add, 12)
+handle.query(:get_value) # => 42
+```
+
+Performing these operations on a child workflow looks very similar:
+
+```ruby
+class ParentWorkflow < Temporal::Workflow
+  def execute
+    # Note that we're using #start_workflow here instead of an #execute_workflow to
+    #   obtain a handle rather than a result
+    handle = workflow.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+
+    # All these calls are blocking, but can be executed in an async block
+    handle.signal(:add, 34)
+    handle.signal(:subtract, 4) # This will send a 'my.signal.subtract' signal as defined
+    handle.signal(:add, 12)
+
+    # Block until workflow finishes
+    handle.result
+  end
+end
+```
+
+
+## Extensibility
+
+To allow extending existing classes with the workflow code without passing workflow context
+explicitly we will expose a module with all the necessary mixins:
+
+```ruby
+require 'temporal/workflow/concern'
+
+class MyClass
+  include Temporal::Workflow::Concern
```

Comment:
Have you instead considered providing something like `DynamicWorkflow` and `DynamicActivity` from other SDKs for extensibility instead? I've found this to be very powerful in Java

Reply:
Those are a different type of extensibility than this.

This provides workflow primitives in non-workflow code so you can build workflow libraries/helpers/mixins.

What you are talking about wrt `DynamicWorkflow`/`DynamicActivity` from Java is the ability to catch unregistered workflows/activities in catch-all implementations. While this is often useful, we haven't built it in many places besides Java, but it'd be fairly easy to build.

Reply:
Agree with Chad that this is a different type of extensibility.

I think that `DynamicWorkflow` and `DynamicActivity` are best implemented using interceptors (aka middleware) — you'd catch a `WorkflowNotFound` / `ActivityNotFound` error and handle it via a dedicated class. This way we don't need a special concept for this use-case and allow for many others.